### PR TITLE
Add generic arch for ARMv7 and update go packages

### PR DESCRIPTION
**Pull request by hgy59**
_Motivation:_  As go build depends on host go only, it is possible to build generic targets for armv7 archs
_Linked issues:_  #3880, #3386

### Checklist
- [x] Build rule `all-supported` completed successfully
- [x] Package upgrade completed successfully
- [x] New installation of package completed successfully

### Remarks
- add arch-armv7 based on armada38x
- update common Makefile with GENERIC_ARCHS to optionally create generic armv7 targets
- update go to version 1.13.8
- update syncthing to version 1.3.4 (includes and closes #3880)
- update minio to latests version
- add dedicated diyspk for fritzctl
- add go build Makefile to implement common compile and install targets
- add go-env Makefile to configure go build (to be used from cross and spk modules)
- use generic armv7 targets in go based packages: fritzctl, minio, syncthing, dnscrypt-proxy, node-exporter

### CODE REVIEW

**Comment by hgy59** [comment]
A similar generic approach may be used for rust builds (#3883)

File: cross/dnscrypt-proxy/Makefile [hunk 1]
```diff
@@ -1,64 +1,29 @@
...
+DEPENDS = native/go
+UNSUPPORTED_ARCHS = $(PPC_ARCHES)
+GENERIC_ARCHS = ARM7
 
 HOMEPAGE = https://github.com/jedisct1/dnscrypt-proxy
 COMMENT  = A flexible DNS proxy, with support for modern encrypted DNS protocols such as DNSCrypt v2 and DNS-over-HTTP/2.
 LICENSE  = ISC
 
-DEPENDS = native/go
+POST_INSTALL_TARGET = dnscrypt_extra_install
```
Comment (ymartin59) [inline cross/dnscrypt-proxy/Makefile]: I find no "extra" here... prefer to name it `dnscrypt_post_install`

File: mk/spksrc.cross-go.mk [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	@for depend in $(DEPENDS) ; \
+	do                          \
+	  $(MAKE) WORK_DIR=$(WORK_DIR) --no-print-directory -C ../../$$depend cat_PLIST ; \
+	done
+	@if [ -f PLIST ] ; \
+	then \
+	  $(PLIST_TRANSFORM) PLIST ; \
+	else \
+	  $(MSG) "No PLIST for $(NAME)" >&2; \
+	fi
```
Comment (ymartin59) [inline mk/spksrc.cross-go.mk]: Sounds to me that section has been copy-paste... Probably a spksrc.plist.mk to be included would be relevant

File: mk/spksrc.cross-go.mk [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	do \
+	  $(MAKE) --no-print-directory -C ../../$$depend dependency-tree ; \
+	done
+
+.PHONY: all-archs
+all-archs: $(addprefix arch-,$(AVAILABLE_ARCHS))
+
+arch-%:
+	@$(MSG) Building package for arch $*
+	-@MAKEFLAGS= $(MAKE) ARCH=$(basename $(subst -,.,$(basename $(subst .,,$*)))) TCVERSION=$(if $(findstring $*,$(basename $(subst -,.,$(basename $(subst .,,$*))))),$(DEFAULT_TC),$(notdir $(subst -,/,$*)))
```
Comment (ymartin59) [inline mk/spksrc.cross-go.mk]: Same remark here... common code should be collected in dedicated makefile if possible and include when needed.

File: native/go/Makefile [hunk 4]
```diff
@@ -12,14 +12,14 @@ COMMENT    = Go is an open source programming language that makes it easy to bui
 LICENSE    =
 
 CONFIGURE_TARGET = nop
-COMPILE_TARGET = myComp
+COMPILE_TARGET = compile-native-go
```
Comment (ymartin59) [inline native/go/Makefile]: Please replace dashes by underscores for consistency

**Comment by hgy59** [comment]
back to WIP, node-exporter doesn't build anymore...

File: mk/spksrc.cross-go.mk (hunk 3 above)
Comment (hgy59) [reply mk/spksrc.cross-go.mk]: Finilly I generalized dependency-tree and digests rules

**Comment by ymartin59** [comment]
Hello @hgy59 May you please rebase to adapt to latest minio version (I merged PR without noticing impact here)

**Comment by hgy59** [comment]
@ymartin59 I resolved the conflicts, but need to rebuild and test.

**Comment by publicarray** [comment]
I just compiled dnscrypt-proxy, works great!

Just some feedback:
Why not upgrade to go 1.14.1?

File: toolchains/syno-armv7-6.1/Makefile [hunk 5]
```diff
@@ -0,0 +1,29 @@
+TC_NAME = syno-armv7
+
+# generic toolchain for arm7hf
+TC_ARCH = armada370 armada375 armada38x alpine armadaxp comcerto2k monaco
```
Comment (ymartin59) [inline toolchains/syno-armv7-6.1/Makefile:4]: @hgy59  alpine4k is lacking from my point of view:
- https://github.com/SynoCommunity/spksrc/wiki/Architecture-per-Synology-model
- https://github.com/SynoCommunity/spksrc/blob/master/toolchains/syno-alpine-6.1/Makefile#L3

**Comment by ymartin59** [comment]
@hgy59 Marvelous. Build is fast and number of published packages have been reduced. I love it !

**Comment by acolomb** [comment]
Why didn't you go for the current Syncthing v1.4.0?

That one would be kind of important since it changes internal database formats and could lead to trouble when Syncthing auto-updates from this SPK version, then gets reinstalled for example, so v1.3.4 sees the database that was already migrated to v1.4 format.  I was about to file a PR for v1.4.0 myself, but after seeing your comment in #3880, I assumed it's already being worked on here.

Unfortunately there wasn't any feedback about [my questions](https://github.com/SynoCommunity/spksrc/pull/3880#issuecomment-585092391) regarding the (incompatible) upstream Syncthing SPK either.  This might become a recurring situation as things currently stand.

**Comment by acolomb** [comment]
Scratch that, there was already a separate PR for Syncthing v1.4.0: #3931 

Please consider upgrading soon.

**Comment by Otiel** [comment]
> Why didn't you go for the current Syncthing v1.4.0?
> 
> That one would be kind of important since it changes internal database formats and could lead to trouble when Syncthing auto-updates from this SPK version, then gets reinstalled for example, so v1.3.4 sees the database that was already migrated to v1.4 format.

Yep. 😞 
I went a bit too fast and upgraded my SynoCommunity-Syncthing package without first checking the commit details - didn't see that it would revert Syncthing to v1.3.4, while my Syncthing was already to v1.4.0.

Now the Synology package seems to crash at start. How can I upgrade Syncthing to v1.4.0 again?

**Comment by acolomb** [comment]
See also #3941.

**Comment by Kaelber** [comment]
I doesnt see this discussion, so I repeat. I´m a user of different Synologys and in the last years I use your Syncthing package. Up to now your packages running very good.

The current version of Syncthing is v1.4.0 and tomorrow v1.4.1 will official release. Until v1.4.0 a new database layout is implemented and the database needs a migration from versions <1.4.0, so also from v1.2.2 or v1.3.4. The new database layout could be a confusion, if a user use the inside updater function and update the system files from the old v1.2.2 to the newest one, today v1.4.1-rc.3. I also doe that.

Now you update your package from v1.2.2 to v1.3.4 and messed with the above facts, because a re-migration of the database is not possible and the package crashes. Next is, why you did not go straight to v1.4.0 is also not clear to me, because of course many of the updater are surely now at least on v1.4.0 or even on v1.4.1-rc.3. The database schema no longer fits. Next is now, the whole package no longer runs after the update.

In any case, you not paying attention. I must therefore deleted the SC package.

**Comment by acolomb** [comment]
I can understand your frustration, this is pretty much the worst possible outcome.

@ymartin59 please upgrade to at least version v1.4.0 as soon as possible, and withdraw the current package from downloads in the meantime.

**Comment by acolomb** [comment]
I've rebased #3931 onto master and am currently making test builds. Will make them available ASAP as a stop-gap solution.

@Kaelber don't touch your package data files in the meantime, it should work again without intervention when you install an SPK of version v1.4.0.

**Comment by Kaelber** [comment]
I urgently needed a solution, so I installed the version of Kastelo (Syncthing-SPK from Syncthing for Synology). However, I will test when the time comes. 

For you Andre, I´m "Andy" in the Syncthing Forum, but it was not possible to choose this name here.

**Comment by ymartin59** [comment]
@Kaelber @acolomb This PR is closed and not related to syncthing. Right? So please report a specific issue in a dedicated issue as a "new issue" in spksrc/issues...
Do you know how many mail notifications from GitHub I receive per day? You are really lucky I catch that one!

**Comment by acolomb** [comment]
@Kaelber Luckily the Kastelo and spksrc versions don't conflict in this case, just be careful in case you are trying to share the config and DB among the two installations.

Unfortunately my test builds are not going well, hitting some unexpected errors. Still investigating. Let's continue discussion in your issue #3941.

**Pull request merged**